drivers/regulator/regulator_pca9420.c

```diff
@@ -136,31 +213,33 @@ static int regulator_set_voltage_offset(const struct device *dev,
 					int32_t min_uv, int32_t max_uv,
 					uint32_t off)
 {
-	const struct regulator_config *config = dev->config;
-	struct regulator_data *data = dev->data;
-	int i = 0;
+	const struct regulator_pca9420_config *config = dev->config;
+	const struct regulator_pca9420_common_config *cconfig = config->parent->config;
+	uint8_t reg_val;
+	int32_t voltage;
 
 	if (!regulator_pca9420_is_supported_voltage(dev, min_uv, max_uv) ||
 		min_uv > max_uv) {
 		return -EINVAL;
 	}
-	/* Find closest supported voltage */
-	while (i < config->num_voltages && min_uv > data->voltages[i].uV) {
-		i++;
-	}
-	if (data->voltages[i].uV > max_uv) {
-		LOG_DBG("Regulator could not satisfy voltage range, too narrow");
-		return -EINVAL;
-	}
-	if (i == config->num_voltages) {
-		LOG_WRN("Regulator could not locate supported voltage,"
-				"but voltage is in range.");
+
+	/* obtain closest value that satisfies minimum (>=) */
+	reg_val = ((min_uv - config->min_uv + PCA9420_VOLTAGE_UV_LSB - 1) /
+		   PCA9420_VOLTAGE_UV_LSB);
+
+	/* check if it leads to a voltage under maximum */
+	voltage = config->min_uv + reg_val * PCA9420_VOLTAGE_UV_LSB;
```

Comment:
This formula won't work correctly without some kind of offset. The MODECFG registers for SW2 and LDO2 on this regulator use two fields, `LSB` and `OFFSET`. The `LSB` field gives a range of voltages to use, and `OFFSET` applies an offset to these voltages.

See SW2 for example: 
![image](https://user-images.githubusercontent.com/34665051/202559958-ec5c773e-9d4a-4565-a629-cd9566c4cdae.png)

Therefore, to select 2.1V output, you should set `SW2_OUT_0_LSB` to any value between `0b11001` and `0b11111`, and clear `SW2_OUT_0_OFFSET`.
No range between 2.1V-2.7V (exclusive) is supported.
To select 2.7V, you should set  `SW2_OUT_0_LSB` to `0b00000`, and set `SW2_OUT_0_OFFSET` to `0b1`.

This limitation is part of the reason I initially used tables to store this data (along with the desire to keep the driver somewhat generic). With a regulator specific driver, we can probably accomplish this without using so much ROM.

Reply:
Thanks for these details! I'll spend some more time reading the datasheet and revisit my changes. I've probably been too optimistic here.

Reply:
No worries, it does appear at first glance that voltage setting is continuous (I think my first pass of the driver might have had this same approach). Just mention me whenever you have an update, and I can test the changes.

Reply:
@danieldegrasse I've added a new linear_range API (as in Linux) to handle such cases, please check latest revision.